### PR TITLE
Add `echo` param to `Console.capture`

### PR DESCRIPTION
**Pull request by darrenburns**
## Type of changes

- [ ] Bug fix
- [x] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [ ] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description

Adds `echo` parameter to the `capture` method of `Console`. When `True`, captured output will also be written to the output file.

### CODE REVIEW

File: rich/console.py [hunk 1]
```diff
@@ -1951,6 +1965,7 @@ def log(
 
     def _check_buffer(self) -> None:
         """Check if the buffer may be rendered. Render it if it can (e.g. Console.quiet is False)
+        be rendered, then clear it.
```
Comment (willmcgugan) [inline rich/console.py:1968]: Docstring doesn't read too well. Can you rephrase?

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Can update the docs to describe the new `echo` param please.

File: rich/console.py [hunk 2]
```diff
@@ -723,6 +727,7 @@ def __init__(
         self._thread_locals = ConsoleThreadLocals(
             theme_stack=ThemeStack(themes.DEFAULT if theme is None else theme)
         )
+        self._buffer_indices_to_render: Set[int] = {self._thread_locals.buffer_index}
```
Comment (willmcgugan) [inline rich/console.py:730]: Don't quite follow what the intent of this is. Handling nested captures? Can you explain to me on Tuesday please.

**Review by willmcgugan — CHANGES_REQUESTED** [review]

File: rich/console.py (hunk 2 above)
Comment (darrenburns) [reply rich/console.py:730]: Previously we only rendered content if the buffer index was 0. This was a hard-coded check. Now we also want to do this if we're inside a capture context (the buffer index is not 0) and where the `echo` arg is `True`, so we record the indices where that's the case.

File: rich/console.py [hunk 3]
```diff
@@ -849,6 +861,8 @@ def end_capture(self) -> str:
         """
         render_result = self._render_buffer(self._buffer)
         del self._buffer[:]
+        if self._buffer_index in self._buffer_indices_to_render:
+            self._buffer_indices_to_render.remove(self._buffer_index)
```
Comment (willmcgugan) [inline rich/console.py:865]:
```suggestion
        self._buffer_indices_to_render.discard(self._buffer_index)
```

File: rich/console.py [hunk 4]
```diff
@@ -1049,7 +1063,7 @@ def bell(self) -> None:
         """Play a 'bell' sound (if supported by the terminal)."""
         self.control(Control.bell())
 
-    def capture(self) -> Capture:
+    def capture(self, echo: bool = False) -> Capture:
```
Comment (willmcgugan) [inline rich/console.py:1066]: Docstring for new arg por favor

File: tests/test_console.py [hunk 5]
```diff
@@ -368,6 +368,19 @@ def test_capture_and_record(capsys):
     assert out == "ABC\n"
 
 
+def test_capture_echo_outputs_captured_content_to_terminal(capsys):
```
Comment (willmcgugan) [inline tests/test_console.py:371]: Could you add a test with more nesting. e..g. a capture inside a capture. Or a simple `with console:` block within a capture.

File: rich/console.py (hunk 2 above)
Comment (willmcgugan) [inline rich/console.py:730]: I think this will break with multiple threads.

You may need to add `_buffer_indices_to_render` to the thread locals.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
A few requests.

File: tests/test_console.py (hunk 5 above)
Comment (darrenburns) [reply tests/test_console.py:371]: Tried this test passes on `master` (it shouldn't), so I'm not sure nested captures work at all:

```python
def test_capture_echo_nested_capture():
    console = Console()
    console.print(1)
    with console.capture() as capture1:
        console.print(2)
        with console.capture() as capture2:
            console.print(3)
    console.print(4)

    assert capture1.get() == ""
    assert capture2.get() == ""
```

Comment (willmcgugan) [reply tests/test_console.py:371]: I'm wondering what the expected behaviour should be.

The outer capture prevents content from getting to the terminal. So the inner capture has nothing to capture? So maybe this is the expected behaviour.

Could you also test this:

```python
with console.capture():
    with console:
        console.print("foo")
```

The with console should make the block atomic, i.e. it saves up all content until it exists the block.

**Comment by nathanrpage97** [comment]
Just wanted to see if `echo` is the right term or would [`tee`](https://en.wikipedia.org/wiki/Tee_(command)) be better?

File: tests/test_console.py (hunk 5 above)
Comment (darrenburns) [reply tests/test_console.py:371]: > The outer capture prevents content from getting to the terminal. So the inner capture has nothing to capture? So maybe this is the expected behaviour.

I'd still expect the `2` to be captured inside `capture1`.

This test passes:

```
def test_capture_echo_nested_capture_console():
    console = Console()
    with console.capture() as capture1:
        console.print(1)
        with console:
            console.print(2)

    assert capture1.get() == "1\n2\n"
```

Comment (darrenburns) [reply tests/test_console.py:371]: Ok, now that I've refreshed my memory with this stuff 😄 ...

This is the behaviour on `master`, which looks wrong:

```python
def test_capture_echo_nested_capture():
    console = Console()
    console.print(1)
    with console.capture() as capture1:
        console.print(2)
        with console.capture() as capture2:
            console.print(3)
    console.print(4)

    assert capture1.get() == ""
    assert capture2.get() == "2\n3\n"
```

My expectation would be that `capture1` captures `2\n3\n` and `capture2` captures `3\n`.

**Comment by darrenburns** [comment]
@nathanrpage97 I agree, I don't think `echo` is the right term either 🤷‍♂️

**Comment by nathanrpage97** [comment]
Yea, I guess it may be a bit esoteric for the general users of the library

**Comment by ssbarnea** [comment]
@darrenburns Any chance to revive this? The changes introduced by 12.5.1 broke capture support in enrich, https://github.com/pycontribs/enrich/issues/40 and temporary I had to pin-down rich as I would need the echo/tee functionality. Thanks.

**Comment by ssbarnea** [comment]
@willmcgugan That PR happens to be on xmas wishlist for quite some time. I wonder if Santa is feeling generous these days ;)

**Comment by ssbarnea** [comment]
@darrenburns any chance to update it? It would be quite cool to see it in.
